### PR TITLE
restapi stop fails if ?t=int not present

### PR DESCRIPTION
**Pull request by brahmaroutu**
Signed-off-by: Srini Brahmaroutu srbrahma@us.ibm.com

### CODE REVIEW

**Comment by LK4D4** [comment]
Pretty sure it closes some issue. ping @runcom

File: integration-cli/docker_api_containers_test.go [hunk 1]
```diff
@@ -859,3 +859,21 @@ func (s *DockerSuite) TestContainerApiRename(c *check.C) {
 		c.Fatalf("Failed to rename container, expected %v, got %v. Container rename API failed", newName, name)
 	}
 }
+
+func (s *DockerSuite) TestPostContainerStop(c *check.C) {
+	runCmd := exec.Command(dockerBinary, "run", "-d", "busybox", "sh")
+	out, _, err := runCommandWithOutput(runCmd)
+	c.Assert(err, check.IsNil)
+
+	containerID := strings.TrimSpace(out)
```
Comment (duglin) [inline integration-cli/docker_api_containers_test.go]: I think you need to add a waitForContainer() to make sure its running before you do the 'stop'.

**Comment by runcom** [comment]
@LK4D4 actually this suffer the same problem as #12679 and this one relates to #12306

**Comment by runcom** [comment]
and btw once address what's pointed out by @duglin I think this is the correct way of handling timeout (with strconv returning err here). Seems ok to me

File: integration-cli/docker_api_containers_test.go [hunk 2]
```diff
@@ -859,3 +859,21 @@ func (s *DockerSuite) TestContainerApiRename(c *check.C) {
 		c.Fatalf("Failed to rename container, expected %v, got %v. Container rename API failed", newName, name)
 	}
 }
+
+func (s *DockerSuite) TestPostContainerStop(c *check.C) {
+	runCmd := exec.Command(dockerBinary, "run", "-d", "busybox", "sh")
```
Comment (LK4D4) [inline integration-cli/docker_api_containers_test.go]: Actually `sh` will exit immediately. You should use `top`.

File: integration-cli/docker_api_containers_test.go (hunk 1 above)
Comment (LK4D4) [inline integration-cli/docker_api_containers_test.go]: I think there is `waitRun` function, which waiting for `Running`.

**Comment by brahmaroutu** [comment]
ParseInt on anything returned from Form.Get is dangerous, not all errors require you to return. The code is addressing missing param t. Test is fixed to use top.

File: api/server/server.go [hunk 3]
```diff
@@ -1080,9 +1080,13 @@ func (s *Server) postContainersStop(eng *engine.Engine, version version.Version,
 		return fmt.Errorf("Missing parameter")
 	}
 
-	seconds, err := strconv.Atoi(r.Form.Get("t"))
-	if err != nil {
-		return err
+	seconds := 0
+	if len(r.Form.Get("t")) > 0 {
+		secs, err := strconv.Atoi(r.Form.Get("t"))
```
Comment (icecrime) [inline api/server/server.go]: Any reason not to directly store to `seconds`?

Comment (brahmaroutu) [inline api/server/server.go]: would you like me to change it like below?
        seconds, err := strconv.Atoi(r.Form.Get("t"))
        if err != nil {
                if err.(*strconv.NumError).Err == strconv.ErrSyntax {
                        seconds = 0
                } else {
                        return err
                }
        }

Comment (brahmaroutu) [inline api/server/server.go]: no need for two if's off course.

**Comment by runcom** [comment]
@brahmaroutu needs rebase

File: integration-cli/docker_api_containers_test.go [hunk 4]
```diff
@@ -859,3 +859,21 @@ func (s *DockerSuite) TestContainerApiRename(c *check.C) {
 		c.Fatalf("Failed to rename container, expected %v, got %v. Container rename API failed", newName, name)
 	}
 }
+
+func (s *DockerSuite) TestPostContainerStop(c *check.C) {
+	runCmd := exec.Command(dockerBinary, "run", "-d", "busybox", "top")
+	out, _, err := runCommandWithOutput(runCmd)
+	c.Assert(err, check.IsNil)
+
+	containerID := strings.TrimSpace(out)
```
Comment (runcom) [inline integration-cli/docker_api_containers_test.go]: as noted before, you should call `waitRun` here before calling stop also

File: api/server/server.go [hunk 5]
```diff
@@ -1071,7 +1071,9 @@ func (s *Server) postContainersStop(eng *engine.Engine, version version.Version,
 	}
 
 	seconds, err := strconv.Atoi(r.Form.Get("t"))
-	if err != nil {
+	if err != nil && err.(*strconv.NumError).Err == strconv.ErrSyntax {
```
Comment (runcom) [inline api/server/server.go]: IMHO we could just ignore error from Atoi cause it's setting seconds to 0 in any case and we ok with this
`seconds, _ := strconv.Atoi(r.Form.Get("t")` is ok  (https://github.com/docker/docker/commit/e41192a3f8cbfbbfecde03f58a3b2be2b1afd836#commitcomment-10831685)

**Comment by brahmaroutu** [comment]
@runcom I believe that I addressed all your comments, please review.

File: integration-cli/docker_api_containers_test.go [hunk 6]
```diff
@@ -1202,3 +1202,23 @@ func (s *DockerSuite) TestContainersApiChunkedEncoding(c *check.C) {
...
+	c.Assert(waitRun(containerID), check.IsNil)
+
+	statusCode, _, err := sockRequest("POST", "/containers/"+containerID+"/stop", nil)
+
+	// 204 No Content is expected, not 200
+	c.Assert(statusCode, check.Equals, http.StatusNoContent)
+	c.Assert(err, check.IsNil)
+
+	state, err := inspectField(containerID, "State.Running")
```
Comment (runcom) [inline integration-cli/docker_api_containers_test.go]: please check error here with c.Assert

**Comment by brahmaroutu** [comment]
@runcom changed as per your suggestion.

**Comment by runcom** [comment]
LGTM if tests ok

**Comment by runcom** [comment]
ping @calavera

**Comment by calavera** [comment]
cool, LGTM :godmode:

**Pull request merged**